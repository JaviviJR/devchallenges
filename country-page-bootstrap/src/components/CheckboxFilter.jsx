import PropTypes from 'prop-types';
import styles from './CheckboxFilter.module.css';
import { useEffect, useState } from 'react';

function CheckboxFilter({ label, filters, handler }) {
    const [selected, setSelected] = useState([]);

    const isChecked = (value) => {
        return selected.includes(value);
    };

    const handleChange = (event) => {
        const { value } = event.target;
        const checked = isChecked(value);
        if (!checked) {
            setSelected([...selected, value]);
        } else {
            setSelected(selected.filter(item => item !== value));
        }
    };

    useEffect(() => {
        handler(selected);
    }, [selected]);

    return (
        <fieldset className={`${styles.checksFilter}`}>
            <legend className={`${styles.label1}`}>{ label }</legend>
            
            { filters.map(filter => (
                <div key={ filter.value } className="form-check">
                    <input 
                        className="form-check-input"
                        type="checkbox"
                        value={ filter.value }
                        id={ filter.value }
                        onChange={ handleChange }
                        checked={ isChecked(filter.value) }
                    />
                    <label 
                        className="form-check-label"
                        htmlFor={ filter.value }
                    >
                        { filter.label }
                    </label>
                </div>
            ))}

        </fieldset>
    );
}

export default CheckboxFilter;

CheckboxFilter.propTypes = {
    label: PropTypes.string.isRequired,
    filters: PropTypes.array.isRequired,
    handler: PropTypes.func.isRequired
};