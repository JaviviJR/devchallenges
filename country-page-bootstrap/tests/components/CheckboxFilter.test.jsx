import { fireEvent, render, screen } from "@testing-library/react"
import CheckboxFilter from "../../src/components/CheckboxFilter"

describe('testing CheckFilter component', () => {
    const filtersTest = [
        { value: 'filter1Value', label: 'filter1Label' },
        { value: 'filter2Value', label: 'filter2Label' }
    ];

    const handlerMock = jest.fn();

    beforeEach(() => {
        jest.clearAllMocks();
    });

    test('should match with snapshot', () => {
        const { container } = render(
            <CheckboxFilter
                label='labelTest'
                filters={ filtersTest }
                handler={ handlerMock }
            />
        );

        expect(container).toMatchSnapshot();
    })

    test('should show all filters', () => {
        render(
            <CheckboxFilter
                label='labelTest'
                filters={ filtersTest }
                handler={ handlerMock }
            />
        );

        const filter1 = screen.getByRole('checkbox', {name: 'filter1Label'});
        const filter2 = screen.getByRole('checkbox', {name: 'filter2Label'});
        expect( filter1 ).toBeInTheDocument();
        expect( filter2 ).toBeInTheDocument();
        expect(filter1.value).toBe('filter1Value');
        expect(filter2.value).toBe('filter2Value');
    });

    test('should handle click', () => {
        render(
            <CheckboxFilter
                label='labelTest'
                filters={ filtersTest }
                handler={ handlerMock }
            />
        );

        const filter1 = screen.getByRole('checkbox', {name: 'filter1Label'});
        fireEvent.click(filter1);
        expect( handlerMock ).toHaveBeenLastCalledWith(['filter1Value']);
    });
})