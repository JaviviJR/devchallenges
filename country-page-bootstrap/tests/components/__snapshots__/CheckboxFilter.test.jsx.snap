// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`testing CheckFilter component should match with snapshot 1`] = `
<div>
  <fieldset
    class="checksFilter"
  >
    <legend
      class="label1"
    >
      labelTest
    </legend>
    <div
      class="form-check"
    >
      <input
        class="form-check-input"
        id="filter1Value"
        type="checkbox"
        value="filter1Value"
      />
      <label
        class="form-check-label"
        for="filter1Value"
      >
        filter1Label
      </label>
    </div>
    <div
      class="form-check"
    >
      <input
        class="form-check-input"
        id="filter2Value"
        type="checkbox"
        value="filter2Value"
      />
      <label
        class="form-check-label"
        for="filter2Value"
      >
        filter2Label
      </label>
    </div>
  </fieldset>
</div>
`;
